File: backend/src/test/java/org/luxons/sevenwonders/test/api/SevenWondersSession.java
package org.luxons.sevenwonders.test.api;

import org.hildan.jackstomp.Channel;
import org.hildan.jackstomp.JackstompSession;
import org.luxons.sevenwonders.actions.ChooseNameAction;
import org.luxons.sevenwonders.actions.CreateGameAction;
import org.luxons.sevenwonders.actions.JoinGameAction;
import org.luxons.sevenwonders.errors.UIError;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

public class SevenWondersSession {

    private final JackstompSession session;

    public SevenWondersSession(JackstompSession session) {
        this.session = session;
    }

    public JackstompSession getJackstompSession() {
        return session;
    }

    public void disconnect() {
        session.disconnect();
    }

    public Channel<UIError> watchErrors() {
        return session.subscribe("/user/queue/errors", UIError.class);
    }

    public ApiPlayer chooseName(String displayName) throws InterruptedException {
        ChooseNameAction action = new ChooseNameAction();
        action.setPlayerName(displayName);
        return session.request(action, ApiPlayer.class, "/app/chooseName", "/user/queue/nameChoice");
    }

    public Channel<ApiLobby[]> watchGames() {
        return session.subscribe("/topic/games", ApiLobby[].class);
    }

    public ApiLobby createGame(String gameName) throws InterruptedException {
        CreateGameAction action = new CreateGameAction();
        action.setGameName(gameName);

        return session.request(action, ApiLobby.class, "/app/lobby/create", "/user/queue/lobby/joined");
    }

    public ApiLobby joinGame(long gameId) throws InterruptedException {
        JoinGameAction action = new JoinGameAction();
        action.setGameId(gameId);

        ApiLobby lobby = session.request(action, ApiLobby.class, "/app/lobby/join", "/user/queue/lobby/joined");
        assertNotNull(lobby);
        assertEquals(gameId, lobby.getId());
        return lobby;
    }

    public Channel<ApiLobby> watchLobbyUpdates(long gameId) {
        return session.subscribe("/topic/lobby/" + gameId + "/updated", ApiLobby.class);
    }

    public Channel<ApiLobby> watchLobbyStart(long gameId) {
        return session.subscribe("/topic/lobby/" + gameId + "/started", ApiLobby.class);
    }

    public void startGame(long gameId) throws InterruptedException {
        String sendDestination = "/app/lobby/startGame";
        String receiveDestination = "/topic/lobby/" + gameId + "/started";
        boolean received = session.request(null, sendDestination, receiveDestination);
        assertTrue(received);
    }

    public void sayReady() {
        session.send("/app/game/sayReady", null);
    }

    public Channel<ApiPlayerTurnInfo> watchTurns() {
        return session.subscribe("/user/queue/game/turn", ApiPlayerTurnInfo.class);
    }
}

File: backend/src/test/kotlin/org/luxons/sevenwonders/test/api/SevenWondersSession.kt
package org.luxons.sevenwonders.test.api

import org.hildan.jackstomp.Channel
import org.hildan.jackstomp.JackstompSession
import org.junit.Assert.*
import org.luxons.sevenwonders.actions.ChooseNameAction
import org.luxons.sevenwonders.actions.CreateGameAction
import org.luxons.sevenwonders.actions.JoinGameAction
import org.luxons.sevenwonders.errors.ErrorDTO

class SevenWondersSession(val jackstompSession: JackstompSession) {

    fun disconnect() {
        jackstompSession.disconnect()
    }

    fun watchErrors(): Channel<ErrorDTO> {
        return jackstompSession.subscribe("/user/queue/errors", ErrorDTO::class.java)
    }

    @Throws(InterruptedException::class)
    fun chooseName(displayName: String): ApiPlayer {
        val action = ChooseNameAction(displayName)
        return jackstompSession.request(action, ApiPlayer::class.java, "/app/chooseName", "/user/queue/nameChoice")
    }

    fun watchGames(): Channel<Array<ApiLobby>> {
        return jackstompSession.subscribe("/topic/games", Array<ApiLobby>::class.java)
    }

    @Throws(InterruptedException::class)
    fun createGame(gameName: String): ApiLobby {
        val action = CreateGameAction(gameName)
        return jackstompSession.request(action, ApiLobby::class.java, "/app/lobby/create", "/user/queue/lobby/joined")
    }

    @Throws(InterruptedException::class)
    fun joinGame(gameId: Long): ApiLobby {
        val action = JoinGameAction(gameId)
        val lobby =
            jackstompSession.request(action, ApiLobby::class.java, "/app/lobby/join", "/user/queue/lobby/joined")
        assertNotNull(lobby)
        assertEquals(gameId, lobby.id)
        return lobby
    }

    fun watchLobbyUpdates(gameId: Long): Channel<ApiLobby> {
        return jackstompSession.subscribe("/topic/lobby/$gameId/updated", ApiLobby::class.java)
    }

    fun watchLobbyStart(gameId: Long): Channel<ApiLobby> {
        return jackstompSession.subscribe("/topic/lobby/$gameId/started", ApiLobby::class.java)
    }

    @Throws(InterruptedException::class)
    fun startGame(gameId: Long) {
        val sendDestination = "/app/lobby/startGame"
        val receiveDestination = "/topic/lobby/$gameId/started"
        val received = jackstompSession.request(null, sendDestination, receiveDestination)
        assertTrue(received)
    }

    fun sayReady() {
        jackstompSession.send("/app/game/sayReady", "")
    }

    fun watchTurns(): Channel<ApiPlayerTurnInfo> {
        return jackstompSession.subscribe("/user/queue/game/turn", ApiPlayerTurnInfo::class.java)
    }
}
